hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/OMRangerBGSync.java

```diff
@@ -0,0 +1,480 @@
+      if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+        executorService.shutdownNow();
+      }
+
+      if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+        LOG.error("Unable to shutdown OM Ranger Background Sync properly.");
+      }
+    } catch (InterruptedException e) {
+      executorService.shutdownNow();
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  private void executeOneRangerSyncCycle() {
+    int attempt = 0;
+    try {
+      // Taking the lock only makes sure that while we are reading
+      // the current Ozone service version, another multitenancy
+      // request is not changing it. We can drop the lock after that.
+      while (!ozoneManager.getMultiTenantManager()
+          .tryAcquireInProgressMtOp(WAIT_MILI)) {
+        sleep(10);
+      }
+      currentOzoneServiceVersionInOMDB = getOmdbRangerServiceVersion();
+      proposedOzoneServiceVersionInOMDB = getRangerServiceVersion();
+      while (currentOzoneServiceVersionInOMDB !=
+          proposedOzoneServiceVersionInOMDB) {
+        if (++attempt > MAX_ATTEMPT) {
+          break;
+        }
+        ozoneManager.getMultiTenantManager().resetInProgressMtOpState();
+        if (!ozoneManager.isLeaderReady()) {
+          return;
+        }
+        LOG.warn("Executing Ranger Sync Cycle.");
+
+        executeOmdbToRangerSync(currentOzoneServiceVersionInOMDB);
+
+        if (currentOzoneServiceVersionInOMDB !=
+            proposedOzoneServiceVersionInOMDB) {
+          // Submit Ratis Request to sync the new ozone service version in OMDB
+          setOmdbRangerServiceVersion(proposedOzoneServiceVersionInOMDB);
+        }
+        while (!ozoneManager.getMultiTenantManager()
+            .tryAcquireInProgressMtOp(WAIT_MILI)) {
+          sleep(10);
+        }
+        currentOzoneServiceVersionInOMDB = proposedOzoneServiceVersionInOMDB;
+        proposedOzoneServiceVersionInOMDB = getRangerServiceVersion();
+      }
+    } catch (Exception e) {
+      LOG.warn("Exception during a Ranger Sync Cycle. " + e.getMessage());
+    } finally {
+      ozoneManager.getMultiTenantManager().resetInProgressMtOpState();
+    }
+  }
+
+  public long getRangerServiceVersion() throws Exception {
+    return authorizer.getCurrentOzoneServiceVersion(ozoneServiceId);
+  }
+
+  public void setOmdbRangerServiceVersion(long version) throws IOException {
+    // OMDB update goes through RATIS
+    ozoneClient.getObjectStore().rangerServiceVersionSync(version);
+  }
+
+  public long getOmdbRangerServiceVersion() {
+    long lastKnownVersion = 0;
+    try {
+      lastKnownVersion =
+          ozoneManager.getMetadataManager().getOmRangerStateTable()
+              .get(OmMetadataManagerImpl.RANGER_OZONE_SERVICE_VERSION_KEY);
+    } catch (Exception ex) {
+      return 0;
+    }
+    return lastKnownVersion;
+  }
+
+  private void executeOmdbToRangerSync(long baseVersion) throws Exception {
+    ++rangerBGSyncCounter;
+    cleanupSyncState();
```

Comment:
We need locking here to prevent undoing in progress requests. Something like:
```
// Get a consistent snapshot of what tenants exist.
lock()
get tenants/policies from Ranger
get tenants/policies from OM
unlock()

for each tenant from OM:
  lock(tenant)
  Update Ranger if needed with policies and their corresponding roles for tenant.
  unlock(tenant)

for each tenant/policy from Ranger:
  lock(tenant)
  delete policy and its roles from Ranger if not in OM
  unlock(tenant)
```
The global lock could be used in place of a tenant specific lock, but this may lead to unnecessary starvation.